File: Colage/Features/AR/ARDiscoveryView.swift
import SwiftUI
import ARKit
import RealityKit

/// AR discovery mode — floating name bubbles in camera view
/// Falls back to a simulated view in the simulator
struct ARDiscoveryView: View {
    @ObservedObject var students: NearbyStudentsViewModel
    @EnvironmentObject var appState: AppState
    @EnvironmentObject var universityService: UniversityService
    @State private var selectedStudent: NearbyStudent?

    var body: some View {
        ZStack {
            #if targetEnvironment(simulator)
            // Simulator fallback — mock AR view
            SimulatedARView(
                students: students.arFilteredStudents,
                themeColor: universityService.currentTheme?.primary ?? ColageColors.primary,
                onStudentTapped: { student in
                    selectedStudent = student
                }
            )
            #else
            // Real device — ARKit session
            LiveARView(
                students: students.arFilteredStudents,
                themeColor: universityService.currentTheme?.primary ?? ColageColors.primary,
                onStudentTapped: { student in
                    selectedStudent = student
                }
            )
            #endif

            // AR overlay UI
            VStack {
                Spacer()

                // AR radius slider
                VStack(spacing: 8) {
                    HStack {
                        Text("AR Range")
                            .font(ColageFonts.caption)
                            .foregroundStyle(ColageColors.textSecondary)
                        Spacer()
                        Text(students.arDistanceFeet.formattedDistance)
                            .font(ColageFonts.monoSmall)
                            .foregroundStyle(universityService.currentTheme?.primary ?? ColageColors.primary)
                    }

                    Slider(
                        value: $students.arMaxDistance,
                        in: 0...1
                    )
                    .tint(universityService.currentTheme?.primary ?? ColageColors.primary)
                }
                .padding(.horizontal, 20)
                .padding(.vertical, 12)
                .background(.ultraThinMaterial)
                .clipShape(RoundedRectangle(cornerRadius: 16))
                .padding(.horizontal, 16)

                // Bottom info bar
                HStack(spacing: 16) {
                    HStack(spacing: 6) {
                        Circle()
                            .fill(ColageColors.online)
                            .frame(width: 8, height: 8)
                        Text("\(students.arFilteredStudents.count) visible")
                            .font(ColageFonts.captionBold)
                            .foregroundStyle(ColageColors.textPrimary)
                    }

                    Spacer()

                    Text("Floor \(appState.currentFloor)")
                        .font(ColageFonts.monoSmall)
                        .foregroundStyle(ColageColors.textSecondary)
                }
                .padding(.horizontal, 20)
                .padding(.vertical, 12)
                .background(.ultraThinMaterial)
                .clipShape(RoundedRectangle(cornerRadius: 16))
                .padding(.horizontal, 16)
                .padding(.bottom, 100)
            }
        }
        .sheet(item: $selectedStudent) { student in
            MiniProfileSheet(student: student)
                .presentationDetents([.fraction(0.35), .large])
                .presentationDragIndicator(.visible)
                .presentationBackgroundInteraction(.enabled)
        }
    }
}

// MARK: - Simulator Fallback

struct SimulatedARView: View {
    let students: [NearbyStudent]
    let themeColor: Color
    let onStudentTapped: (NearbyStudent) -> Void

    @State private var bubbleOffsets: [String: CGPoint] = [:]

    var body: some View {
        GeometryReader { geo in
            ZStack {
                // Dark background simulating camera
                LinearGradient(
                    colors: [Color(hex: "1a1a2e"), Color(hex: "0a0a1a")],
                    startPoint: .top,
                    endPoint: .bottom
                )
                .ignoresSafeArea()

                // Grid lines for AR effect
                ForEach(0..<8, id: \.self) { i in
                    Rectangle()
                        .fill(Color.white.opacity(0.03))
                        .frame(height: 1)
                        .offset(y: CGFloat(i) * geo.size.height / 8 - geo.size.height / 2)
                }
                ForEach(0..<6, id: \.self) { i in
                    Rectangle()
                        .fill(Color.white.opacity(0.03))
                        .frame(width: 1)
                        .offset(x: CGFloat(i) * geo.size.width / 6 - geo.size.width / 2)
                }

                // "AR" label
                VStack {
                    HStack {
                        Label("SIMULATOR", systemImage: "camera.viewfinder")
                            .font(ColageFonts.captionBold)
                            .foregroundStyle(ColageColors.textTertiary)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(.ultraThinMaterial)
                            .clipShape(Capsule())
                        Spacer()
                    }
                    .padding(.leading, 16)
                    .padding(.top, 90)
                    Spacer()
                }

                // Floating bubbles
                ForEach(students.prefix(10)) { student in
                    let offset = bubbleOffset(for: student, in: geo.size)
                    ARBubble(
                        student: student,
                        themeColor: themeColor
                    )
                    .position(x: offset.x, y: offset.y)
                    .onTapGesture {
                        onStudentTapped(student)
                    }
                    .animation(
                        .easeInOut(duration: 2.0 + Double.random(in: 0...1))
                        .repeatForever(autoreverses: true),
                        value: offset
                    )
                }
            }
        }
        .onAppear {
            generateOffsets()
        }
    }

    private func bubbleOffset(for student: NearbyStudent, in size: CGSize) -> CGPoint {
        if let offset = bubbleOffsets[student.id] {
            return offset
        }
        return CGPoint(x: size.width / 2, y: size.height / 2)
    }

    private func generateOffsets() {
        for student in students.prefix(10) {
            let x = CGFloat.random(in: 60...320)
            let y = CGFloat.random(in: 140...700)
            bubbleOffsets[student.id] = CGPoint(x: x, y: y)
        }
    }
}

// MARK: - AR Bubble

struct ARBubble: View {
    let student: NearbyStudent
    let themeColor: Color

    var body: some View {
        VStack(spacing: 4) {
            // Avatar circle
            AvatarView(
                imageURL: student.profile.profilePhotoURL,
                size: 48,
                borderColor: themeColor,
                initials: student.profile.displayName.initials
            )
            .shadow(color: themeColor.opacity(0.4), radius: 8)

            // Name tag
            VStack(spacing: 2) {
                Text(student.profile.displayName.components(separatedBy: " ").first ?? "")
                    .font(ColageFonts.captionBold)
                    .foregroundStyle(.white)

                Text(student.distance.formattedDistance)
                    .font(ColageFonts.monoSmall)
                    .foregroundStyle(themeColor)
            }
            .padding(.horizontal, 10)
            .padding(.vertical, 6)
            .background(.ultraThinMaterial)
            .clipShape(RoundedRectangle(cornerRadius: 10))
        }
    }
}

// MARK: - Live AR View (real device only)

#if !targetEnvironment(simulator)
struct LiveARView: UIViewRepresentable {
    let students: [NearbyStudent]
    let themeColor: Color
    var onStudentTapped: ((NearbyStudent) -> Void)?

    func makeCoordinator() -> Coordinator {
        Coordinator(parent: self)
    }

    func makeUIView(context: Context) -> ARView {
        let arView = ARView(frame: .zero)

        // World tracking with camera feed
        let config = ARWorldTrackingConfiguration()
        config.planeDetection = []
        arView.session.run(config)

        context.coordinator.arView = arView
        context.coordinator.addStudentEntities(students: students)

        // Tap gesture
        let tap = UITapGestureRecognizer(target: context.coordinator, action: #selector(Coordinator.handleTap(_:)))
        arView.addGestureRecognizer(tap)

        return arView
    }

    func updateUIView(_ arView: ARView, context: Context) {
        context.coordinator.updateStudentEntities(students: students)
    }

    class Coordinator: NSObject {
        let parent: LiveARView
        var arView: ARView?
        private var anchors: [String: AnchorEntity] = [:]

        init(parent: LiveARView) {
            self.parent = parent
        }

        func addStudentEntities(students: [NearbyStudent]) {
            guard let arView = arView else { return }

            for student in students.prefix(15) {
                let anchor = AnchorEntity(world: randomPosition(distance: student.distance))

                // Simple sphere for now — custom meshes later
                let mesh = MeshResource.generateSphere(radius: 0.08)
                let material = SimpleMaterial(color: UIColor(red: 0.647, green: 0.110, blue: 0.188, alpha: 1.0), isMetallic: false) // Crimson
                let entity = ModelEntity(mesh: mesh, materials: [material])
                entity.name = student.id

                anchor.addChild(entity)
                arView.scene.addAnchor(anchor)
                anchors[student.id] = anchor
            }
        }

        func updateStudentEntities(students: [NearbyStudent]) {
            // Update positions based on new data
        }

        @objc func handleTap(_ gesture: UITapGestureRecognizer) {
            guard let arView = arView else { return }
            let location = gesture.location(in: arView)

            if let entity = arView.entity(at: location) {
                let studentId = entity.name
                if let student = parent.students.first(where: { $0.id == studentId }) {
                    parent.onStudentTapped?(student)
                }
            }
        }

        private func randomPosition(distance: Double) -> SIMD3<Float> {
            let meters = Float(distance) * 0.3048 // feet to meters
            let angle = Float.random(in: 0...(2 * .pi))
            let height = Float.random(in: -0.5...1.5)
            let clampedDist = min(meters, 10.0) // Cap at 10m for visibility
            return SIMD3(
                cos(angle) * clampedDist,
                height,
                -sin(angle) * clampedDist
            )
        }
    }
}
#endif

#Preview {
    ARDiscoveryView(students: {
        let vm = NearbyStudentsViewModel()
        vm.loadMockData()
        return vm
    }())
    .environmentObject(AppState())
    .environmentObject(UniversityService())
}

File: android/app/src/main/java/com/colageclub/colage/features/ar/ARDiscoveryView.kt
package com.colageclub.colage.features.ar

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.colageclub.colage.core.design.*
import com.colageclub.colage.data.models.NearbyStudent
import com.colageclub.colage.features.discovery.MiniProfileSheet
import com.colageclub.colage.features.discovery.NearbyStudentsViewModel
import kotlin.random.Random

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ARDiscoveryView(
    viewModel: NearbyStudentsViewModel,
    currentFloor: Int,
    themeColor: Color = LocalThemeColor.current
) {
    val arMaxDistance by viewModel.arMaxDistance.collectAsState()
    val students = viewModel.arFilteredStudents()
    var selectedStudent by remember { mutableStateOf<NearbyStudent?>(null) }

    Box(modifier = Modifier.fillMaxSize()) {
        // Simulated AR background (no ARCore in emulator)
        SimulatedARBackground(
            students = students.take(10),
            themeColor = themeColor,
            onStudentTapped = { selectedStudent = it }
        )

        // Overlay UI
        Column(
            modifier = Modifier.fillMaxSize(),
            verticalArrangement = Arrangement.Bottom
        ) {
            // AR Range slider
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(ColageColors.Surface.copy(alpha = 0.85f))
                    .padding(horizontal = 20.dp, vertical = 12.dp)
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text("AR Range", style = ColageFonts.Caption.copy(color = ColageColors.TextSecondary))
                    Text(
                        viewModel.arDistanceFeet.formattedDistance(),
                        style = ColageFonts.MonoSmall.copy(color = themeColor)
                    )
                }
                Slider(
                    value = arMaxDistance,
                    onValueChange = { viewModel.setArMaxDistance(it) },
                    valueRange = 0f..1f,
                    colors = SliderDefaults.colors(thumbColor = themeColor, activeTrackColor = themeColor)
                )
            }

            Spacer(Modifier.height(8.dp))

            // Bottom info bar
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(ColageColors.Surface.copy(alpha = 0.85f))
                    .padding(horizontal = 20.dp, vertical = 12.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                    Box(
                        modifier = Modifier
                            .size(8.dp)
                            .clip(CircleShape)
                            .background(ColageColors.Online)
                    )
                    Text("${students.size} visible", style = ColageFonts.CaptionBold.copy(color = ColageColors.TextPrimary))
                }
                Text("Floor $currentFloor", style = ColageFonts.MonoSmall.copy(color = ColageColors.TextSecondary))
            }

            Spacer(Modifier.height(100.dp))
        }
    }

    // Bottom sheet
    selectedStudent?.let { student ->
        val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)
        ModalBottomSheet(
            onDismissRequest = { selectedStudent = null },
            sheetState = sheetState,
            containerColor = ColageColors.Background,
            dragHandle = { BottomSheetDefaults.DragHandle(color = ColageColors.Border) }
        ) {
            MiniProfileSheet(student = student, themeColor = themeColor)
        }
    }
}

@Composable
fun SimulatedARBackground(
    students: List<NearbyStudent>,
    themeColor: Color,
    onStudentTapped: (NearbyStudent) -> Unit
) {
    // Generate stable random positions
    val positions = remember(students.map { it.id }) {
        students.associate { student ->
            student.id to Pair(
                Random.nextFloat() * 0.7f + 0.1f, // x: 10-80% of width
                Random.nextFloat() * 0.5f + 0.15f  // y: 15-65% of height
            )
        }
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(
                Brush.verticalGradient(
                    colors = listOf(Color(0xFF1A1A2E), Color(0xFF0A0A1A))
                )
            )
    ) {
        // Grid lines for AR effect
        repeat(8) { i ->
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(1.dp)
                    .offset(y = (i * 100).dp)
                    .background(Color.White.copy(alpha = 0.03f))
            )
        }

        // SIMULATOR label
        Row(
            modifier = Modifier
                .padding(start = 16.dp, top = 90.dp)
                .background(ColageColors.Surface.copy(alpha = 0.6f), RoundedCornerShape(50))
                .padding(horizontal = 12.dp, vertical = 6.dp)
        ) {
            Icon(Icons.Default.CameraAlt, null, tint = ColageColors.TextTertiary, modifier = Modifier.size(12.dp))
            Spacer(Modifier.width(4.dp))
            Text("SIMULATOR", style = ColageFonts.CaptionBold.copy(color = ColageColors.TextTertiary))
        }

        // Floating bubbles
        BoxWithConstraints(modifier = Modifier.fillMaxSize()) {
            students.forEach { student ->
                val pos = positions[student.id] ?: return@forEach
                val x = maxWidth * pos.first
                val y = maxHeight * pos.second

                ARBubble(
                    student = student,
                    themeColor = themeColor,
                    modifier = Modifier
                        .offset(x = x, y = y)
                        .clickable { onStudentTapped(student) }
                )
            }
        }
    }
}

@Composable
fun ARBubble(
    student: NearbyStudent,
    themeColor: Color,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        AvatarView(
            imageUrl = student.profile.profilePhotoURL,
            size = 48.dp,
            borderColor = themeColor,
            initials = student.profile.displayName.initials()
        )

        Spacer(Modifier.height(4.dp))

        Column(
            modifier = Modifier
                .background(ColageColors.Surface.copy(alpha = 0.8f), RoundedCornerShape(10.dp))
                .padding(horizontal = 10.dp, vertical = 6.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                student.profile.displayName.split(" ").first(),
                style = ColageFonts.CaptionBold.copy(color = Color.White)
            )
            Text(
                student.distance.formattedDistance(),
                style = ColageFonts.MonoSmall.copy(color = themeColor)
            )
        }
    }
}
